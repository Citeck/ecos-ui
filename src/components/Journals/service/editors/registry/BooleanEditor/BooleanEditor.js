import React, { useCallback, useEffect, useState } from 'react';
import isFunction from 'lodash/isFunction';
import get from 'lodash/get';

import { t } from '../../../../../../helpers/export/util';
import { getBool } from '../../../../../../helpers/util';
import ZIndex from '../../../../../../services/ZIndex';
import { Checkbox, Select } from '../../../../../common/form';
import EditorScope from '../../EditorScope';
import BaseEditor from '../BaseEditor';

const Modes = { select: 'select', checkbox: 'checkbox' };

export default class BooleanEditor extends BaseEditor {
  static TYPE = 'boolean';

  static options = [
    {
      value: null,
      get label() {
        return t('react-select.select-value.label');
      }
    },
    {
      value: true,
      get label() {
        return t('react-select.value-true.label');
      }
    },
    {
      value: false,
      get label() {
        return t('react-select.value-false.label');
      }
    }
  ];

  getControl(config, scope) {
    return ({ value, onUpdate }) => {
      const [selected, setSelected] = useState(value);
      const [zIndex, setZIndex] = useState(ZIndex.calcZ() + 1);
      const mode = config.mode || Modes.select;
      const _value = getBool(value);

      if (mode === Modes.checkbox) {
        return <Checkbox className="p-1" checked={_value} onChange={e => onUpdate(e.checked)} />;
      }

      useEffect(() => {
        const newZIndex = ZIndex.calcZ() + 1;
        if (zIndex !== newZIndex) {
          setZIndex(newZIndex);
        }
      });

      useEffect(() => {
        const selected = BooleanEditor.options.find(opt => get(opt, 'value', opt) === value) || null;

        setSelected(selected);
      }, [value]);

      const onSelect = useCallback(
        selected => {
          if (isFunction(onUpdate)) {
            onUpdate(selected.value);
          }

          setSelected(selected);
        },
        [value]
      );

      return (
        <Select
          options={BooleanEditor.options}
          value={selected}
          autoFocus={scope === EditorScope.CELL}
          onChange={onSelect}
          isSearchable={false}
          className="select_narrow select_width_full ecosZIndexAnchor"
          menuPortalTarget={document.body}
          menuPlacement="auto"
          closeMenuOnScroll={(e, { innerSelect }) => !innerSelect}
          styles={{ menuPortal: base => ({ ...base, zIndex }) }}
        />
      );
    };
  }
}
