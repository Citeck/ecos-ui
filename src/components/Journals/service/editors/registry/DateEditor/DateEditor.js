import React from 'react';

import BaseEditor from '../BaseEditor';
import DateEditorControl from './DateEditorControl';

export default class DateEditor extends BaseEditor {
  static TYPE = 'date';

  getControl(config, scope) {
    return props => <DateEditorControl scope={scope} {...props} />;
  }
}
