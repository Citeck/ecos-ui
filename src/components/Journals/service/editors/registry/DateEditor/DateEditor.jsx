import BaseEditor from '../BaseEditor';
import DateEditorView from './DateEditorView';

export default class DateEditor extends BaseEditor {
  static TYPE = 'date';

  get viewComponent() {
    return DateEditorView;
  }
}
