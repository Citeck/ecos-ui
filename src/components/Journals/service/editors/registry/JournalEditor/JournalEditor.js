import React from 'react';

import Records from '../../../../../Records/Records';
import SelectJournal from '../../../../../common/form/SelectJournal';
import EditorScope from '../../EditorScope';
import BaseEditor from '../BaseEditor';

export default class JournalEditor extends BaseEditor {
  static TYPE = 'journal';

  getControl(config, scope) {
    const journalId = config.journalId;

    return ({ value, onUpdate, onCancel, multiple }) => (
      <SelectJournal
        multiple={multiple}
        autoFocus={scope === EditorScope.CELL}
        isCompact
        inputViewClass="select-journal__input-view_extra-compact"
        journalId={journalId}
        defaultValue={value}
        onChange={onUpdate}
        onCancel={onCancel}
      />
    );
  }

  getDisplayName(value, config, scope, state) {
    return Records.get(value).load('?disp');
  }
}
