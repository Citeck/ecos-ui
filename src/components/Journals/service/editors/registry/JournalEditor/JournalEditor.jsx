import BaseEditor from '../BaseEditor';
import JournalEditorView from './JournalEditorView';

export default class JournalEditor extends BaseEditor {
  static TYPE = 'journal';

  get viewComponent() {
    return JournalEditorView;
  }

  getRecordValue(record) {
    return record.value;
  }
}
