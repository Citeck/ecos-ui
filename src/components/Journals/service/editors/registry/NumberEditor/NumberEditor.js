import TextEditor from '../TextEditor';

export default class NumberEditor extends TextEditor {
  static TYPE = 'number';
  inputType = NumberEditor.TYPE;
}
