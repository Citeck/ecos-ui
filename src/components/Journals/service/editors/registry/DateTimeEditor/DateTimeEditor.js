import React from 'react';

import DateEditor from '../DateEditor';

import DateTimeEditorControl from './DateTimeEditorControl';

export default class DateTimeEditor extends DateEditor {
  static TYPE = 'datetime';

  getControl(config, scope, params) {
    return props => <DateTimeEditorControl scope={scope} {...props} {...params} />;
  }
}
