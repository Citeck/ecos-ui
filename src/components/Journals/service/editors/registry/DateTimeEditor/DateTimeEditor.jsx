import DateEditor from '../DateEditor';
import DateTimeEditorView from './DateTimeEditorView';

export default class DateTimeEditor extends DateEditor {
  static TYPE = 'datetime';

  get viewComponent() {
    return DateTimeEditorView;
  }
}
