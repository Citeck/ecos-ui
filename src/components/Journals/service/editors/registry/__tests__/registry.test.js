/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, screen } from '@testing-library/react';

import editorRegistry from '../';
import BooleanEditor from '../BooleanEditor';
import DateEditor from '../DateEditor';
import DateTimeEditor from '../DateTimeEditor';
import JournalEditor from '../JournalEditor';
import NumberEditor from '../NumberEditor';
import OrgstructEditor from '../OrgstructEditor';
import SelectEditor from '../SelectEditor';
import TextEditor from '../TextEditor';
import { DatePicker } from '../../../../../common/form';

const WRAPPER = true;

function getEditorComponent({ type, config, value, wrapper }) {
  const editorInstance = editorRegistry.getEditor(type);
  const Control = editorInstance.getControl(config, {});

  return render(<Control config={config} value={value} />);
}

describe('editors registry', () => {
  describe.each([
    ['boolean', BooleanEditor],
    ['date', DateEditor],
    ['datetime', DateTimeEditor],
    ['journal', JournalEditor],
    ['number', NumberEditor],
    ['orgstruct', OrgstructEditor],
    ['select', SelectEditor],
    ['text', TextEditor],
    [undefined, TextEditor],
    ['unknown', TextEditor],
  ])('type > editor class', (type, editorClass) => {
    test(`type "${type}" expects editor "${editorClass.name}"`, () => {
      const editorInstance = editorRegistry.getEditor(type);
      expect(editorInstance instanceof editorClass).toBeTruthy();
    });
  });

  describe.each([
    ['boolean', 'ecos-select'],
    ['boolean', 'ecos-select', { mode: 'select' }],
    ['boolean', 'ecos-checkbox', { mode: 'checkbox' }],
    ['date', 'ecos-datepicker', {}, WRAPPER],
    ['datetime', 'ecos-datepicker', {}, WRAPPER],
    ['journal', 'select-journal'],
    ['orgstruct', 'select-orgstruct', {}, WRAPPER],
    // ['select', 'div', {options: []}], // TODO: isLoading state incorrect
    ['number', 'ecos-input'],
    ['text', 'ecos-input'],
    [undefined, 'ecos-input'],
    ['unknown', 'ecos-input'],
  ])('type > expected element', (type, controlClassName, config = {}, wrapper) => {
    const editorInstance = editorRegistry.getEditor(type);
    const Control = editorInstance.getControl(config, {});

    test(`type "${type}" has element by class=${controlClassName}`, () => {
      const { container } = render(<Control config={config} value={undefined} />);

      expect(container.getElementsByClassName(controlClassName)).toHaveLength(1);
    });
  });

  describe.each([
    ['boolean', { config: { mode: 'checkbox' }, value: true, checked: true }],
    ['boolean', { config: { mode: 'checkbox' }, value: false, checked: false }],
    ['date', { showTimeInput: undefined, dateFormat: 'dd.MM.yyyy' }, WRAPPER],
    ['datetime', { showTimeInput: true, dateFormat: 'dd.MM.yyyy HH:mm' }, WRAPPER],
    ['journal', { config: { journalId: '1' }, journalId: '1' }],
    ['orgstruct', { config: { allowedAuthorityTypes: 'USER' }, allowedAuthorityTypes: ['USER'] }, WRAPPER],
    ['number', { type: 'number' }],
    ['text', { type: 'text' }],
  ])('type > expected props', (type, params = {}, wrapper) => {
    const { config = {}, value, ...props } = params;
    const editorInstance = editorRegistry.getEditor(type);
    const Control = editorInstance.getControl(config, {});

    const { asFragment } = getEditorComponent({ type, config, value, wrapper });

    test(`type "${type}" expects props`, () => {
      expect(asFragment(<Control {...props} />)).toMatchSnapshot();
    });
  });
});
