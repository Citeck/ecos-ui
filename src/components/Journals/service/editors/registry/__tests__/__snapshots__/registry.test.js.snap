// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editors registry type > expected props type "boolean" expects props 1`] = `<DocumentFragment />`;

exports[`editors registry type > expected props type "boolean" expects props 2`] = `<DocumentFragment />`;

exports[`editors registry type > expected props type "date" expects props 1`] = `<DocumentFragment />`;

exports[`editors registry type > expected props type "datetime" expects props 1`] = `<DocumentFragment />`;

exports[`editors registry type > expected props type "journal" expects props 1`] = `<DocumentFragment />`;

exports[`editors registry type > expected props type "number" expects props 1`] = `<DocumentFragment />`;

exports[`editors registry type > expected props type "orgstruct" expects props 1`] = `<DocumentFragment />`;

exports[`editors registry type > expected props type "text" expects props 1`] = `<DocumentFragment />`;
